source/extensions/filters/listener/http_inspector/http_inspector.cc

```diff
@@ -64,47 +65,98 @@ void Filter::onRead() {
-  if (absl::StartsWith(data, Filter::HTTP2_CONNECTION_PREFACE)) {
+  size_t len = std::min(data.length(), Filter::HTTP2_CONNECTION_PREFACE.length());
+  if (Filter::HTTP2_CONNECTION_PREFACE.compare(0, len, data, 0, len) == 0) {
+    if (data.length() < Filter::HTTP2_CONNECTION_PREFACE.length()) {
+      return;
+    }
     ENVOY_LOG(trace, "http inspector: http2 connection preface found");
     protocol_ = "HTTP/2";
     done(true);
   } else {
     const size_t pos = data.find_first_of("\r\n");
+    if (pos != absl::string_view::npos) {
+      const absl::string_view request_line = data.substr(0, pos);
+      std::vector<absl::string_view> fields = absl::StrSplit(request_line, absl::MaxSplits(' ', 4));
+
+      // Method SP Request-URI SP HTTP-Version
+      if (fields.size() != 3) {
+        ENVOY_LOG(trace, "http inspector: invalid http1x request line");
+        return done(false);
+      }
+
+      if (httpMethods().count(fields[0]) == 0 || httpProtocols().count(fields[2]) == 0) {
+        ENVOY_LOG(trace, "http inspector: method: {} or protocol: {} not valid", fields[0],
+                  fields[2]);
+        return done(false);
+      }
+
+      ENVOY_LOG(trace, "http inspector: method: {}, request uri: {}, protocol: {}", fields[0],
+                fields[1], fields[2]);
+
+      protocol_ = fields[2];
+      return done(true);
+    }
 
     // Cannot find \r or \n
-    if (pos == absl::string_view::npos) {
-      ENVOY_LOG(trace, "http inspector: no request line detected");
+    ENVOY_LOG(trace, "http inspector: no request line detected");
+    std::vector<absl::string_view> fields = absl::StrSplit(data, absl::MaxSplits(' ', 4));
+
+    // Check if inspect partial of HTTP 1.x packet.
+    switch (fields.size()) {
+    case 0:
+      break;
+    case 1:
+      // Check if partial of HTTP method is received.
+      if (!checkPrefix(fields[0], httpMethods())) {
+        return done(false);
+      }
+      break;
+    case 2:
+      /**
+       * e.g. GET /index -> waiting for following bytes
+       *      BAD /index -> exit inspector
+       */
+      if (httpMethods().count(fields[0]) == 0) {
+        return done(false);
+      }
+      break;
+    case 3:
+      /**
+       * e.g. GET /index HTTP -> waiting for following bytes
+       *      GET /index BAD -> exit inspector
+       *      BAD /index HTTP/1.1 -> exit inspector
+       */
+      if (httpMethods().count(fields[0]) == 0 || !checkPrefix(fields[2], httpProtocols())) {
+        return done(false);
+      }
+      break;
+    default:
       return done(false);
     }
+  }
+}
 
-    const absl::string_view request_line = data.substr(0, pos);
-    std::vector<absl::string_view> fields = absl::StrSplit(request_line, absl::MaxSplits(' ', 4));
-
-    // Method SP Request-URI SP HTTP-Version
-    if (fields.size() != 3) {
-      ENVOY_LOG(trace, "http inspector: invalid http1x request line");
-      return done(false);
+bool Filter::checkPrefix(absl::string_view prefix,
```

Comment:
@PiotrSikora @lizan Do you think we need to use more efficient data structures(e.g. trie) to accelerate the query?